src/charm.py

```diff
@@ -100,23 +98,47 @@ def _on_config_changed(self, event):
 
     def publish_config(self):
         """Generate config file and publish to all workers."""
-        mimir_config = self.coordinator.build_config(dict(self.config))
+        mimir_config = self.coordinator.build_config(self._s3_storage_data)
         self.cluster_provider.publish_configs(mimir_config)
 
+    def _s3_connection_info(self):
+        """Parse a _S3StorageBackend object from relation data."""
+        if not self.s3_requirer.relations:
+            self._s3_storage_data = _S3StorageBackend()
+            return
+        raw = self.s3_requirer.get_s3_connection_info()
+        raw["access_key"], raw["secret_key"] = raw.pop("access-key", ""), raw.pop("secret-key", "")
+        self._s3_storage_data = _S3StorageBackend(**raw)
+
     def _on_mimir_cluster_changed(self, _):
-        if self.coordinator.is_coherent():
-            logger.info("mimir deployment coherent: publishing configs")
-            self.publish_config()
-        else:
-            logger.warning("this mimir deployment is incoherent")
+        if not self.coordinator.is_coherent():
+            logger.warning("Incoherent deployment: Some required Mimir roles are missing.")
+            return
+        self._s3_connection_info()
+        if self._s3_storage_data == _S3StorageBackend() and self.coordinator.is_scaled():
```

Comment:
here I'd do something like add a property `self._is_s3_available` to check that we have a s3 relation and we have a bucket configured, instead of relying on that instance attr